src/igmp.c

```diff
@@ -282,14 +289,24 @@ static void buildIgmp(uint32_t src, uint32_t dst, int type, int code, uint32_t g
     ((unsigned char*)send_buf+MIN_IP_HEADER_LEN)[2] = 0x00;
     ((unsigned char*)send_buf+MIN_IP_HEADER_LEN)[3] = 0x00;
 
-    igmp                    = (struct igmp *)(send_buf + IP_HEADER_RAOPT_LEN);
+    igmp                    = (struct igmpv3 *)(send_buf + IP_HEADER_RAOPT_LEN);
     igmp->igmp_type         = type;
     igmp->igmp_code         = code;
     igmp->igmp_group.s_addr = group;
     igmp->igmp_cksum        = 0;
-    igmp->igmp_cksum        = inetChksum((unsigned short *)igmp,
-                                         IP_HEADER_RAOPT_LEN + datalen);
 
+    if(conf->useIgmpv3) {
+        igmp->igmp_misc         = conf->robustnessValue & 0x7;
+        igmp->igmp_qqi          = conf->queryInterval;
+        igmp->igmp_numsrc       = 0;
+        igmp->igmp_cksum        = inetChksum((unsigned short *)igmp,
+                                            IGMP_V3_QUERY_MINLEN + datalen);
+    } else {
+        igmp->igmp_cksum        = inetChksum((unsigned short *)igmp,
+                                            IP_HEADER_RAOPT_LEN + datalen);
+    }
```

Comment:
You have stored query length in the `query_minlen` variable. So you can move `inetChksum` code outside of the `if` and call it based on the `query_minlen`.